wremnants/include/muon_efficiencies_binned_vqt_real.h

```diff
@@ -0,0 +1,475 @@
+            auto const eff_type_idx_iso_pass = with_trigger ? idx_iso_triggering_ : idx_iso_nontriggering_;
+            auto const eff_type_idx_iso_fail = with_trigger ? idx_antiiso_triggering_ : idx_antiiso_nontriggering_;
+            auto const eff_type_idx_iso = pass_iso ? eff_type_idx_iso_pass : eff_type_idx_iso_fail;
+
+            const double reco      = sf_reco_->at(eta_idx, pt_idx_reco, charge_idx, eff_type_idx_reco,      idx_nom_alt).value();
+            const double tracking  = sf_tracking_->at(saeta_idx, sapt_idx,    charge_idx, eff_type_idx_tracking,  idx_nom_alt).value();
+            const double idip      = sf_idip_trig_iso_->at(  eta_idx,   pt_idx,    charge_idx, eff_type_idx_idip,      idx_nom_alt).value();
+            double trig = 1.0;
+            if (with_trigger) trig = sf_idip_trig_iso_->at(  eta_idx,   pt_idx,    charge_idx, eff_type_idx_trig,      idx_nom_alt).value();
+            const double iso       = sf_idip_trig_iso_->at(  eta_idx,   pt_idx,    charge_idx, eff_type_idx_iso,       idx_nom_alt).value();
+
+            std::array<double,5> ret = {reco, tracking, idip, trig, iso};
+            // for(int i = 0; i < ret.size(); i++)
+            //     std::cout << "Scale factor i = " << i << " --> " << ret[i] << std::endl;
+
+            return ret;
+            
+        }
+
+        double scale_factor_product(float pt, float eta, float sapt, float saeta, int charge, bool pass_iso, bool with_trigger, float vqt, int idx_nom_alt) const {
+
+            auto const eta_idx = sf_idip_trig_iso_->template axis<0>().index(eta);
+            auto const pt_idx  = sf_idip_trig_iso_->template axis<1>().index(pt);
+            auto const charge_idx = sf_idip_trig_iso_->template axis<2>().index(charge);
+            auto const pt_idx_reco = sf_reco_->template axis<1>().index(pt);
+            auto const saeta_idx   = sf_tracking_->template axis<0>().index(saeta);
+            auto const sapt_idx    = sf_tracking_->template axis<1>().index(sapt);
+
+            std::array<double,5> allSF = scale_factor_array(pt_idx, pt_idx_reco, sapt_idx, eta_idx, saeta_idx, charge_idx, pass_iso, with_trigger, idx_nom_alt);
+            double sf = 1.0;
+            if (includeTrigger_) {
+              for(int i = 0; i < (allSF.size()-2); i++) {
+                  // std::cout << "Scale factor i = " << i << " --> " << allSF[i] << std::endl;
+                  sf *= allSF[i];
+              }
+            }
+            else {
+              for(int i = 0; i < (allSF.size()-1); i++) {
+                  // std::cout << "Scale factor i = " << i << " --> " << allSF[i] << std::endl;
+                  sf *= allSF[i];
+              }
+            }
+            int chargeidx=(1+charge)/2;
+            int bin=iso3d->FindBin(vqt,eta,pt), binlow=iso3d->FindBin(iso3d->GetXaxis()->GetBinCenter(1),eta,pt), binup=iso3d->FindBin(iso3d->GetXaxis()->GetBinCenter(iso3d->GetXaxis()->GetNbins()),eta,pt);
```

Comment:
Usually I try to use FindFixBin instead of FindBin: https://root.cern.ch/doc/v608/classTH1.html#aee13ee34b23f0a238af6279ef5441bc4

They are practically equivalent, but in case one is reading the overflows FindBin performs additional operations which might be slower (although I am not sure by how much to be honest)

Also, sometimes operations like h->GetOneAxis->GetBinSomething are done multiple times on the same bin, it might be good to store the value in a variable to reuse it where needed, it would also make the code cleaner by shortening the length of some lines.